File: python/robot-simulator/robot_simulator.py
class NORTH:
    @staticmethod
    def turn_right(self):
        return EAST

    @staticmethod
    def turn_left(self):
        return WEST

    @staticmethod
    def advance(self, x, y):
        return (x, y + 1)

class EAST:
    @staticmethod
    def turn_right(self):
        return SOUTH

    @staticmethod
    def turn_left(self):
        return NORTH

    @staticmethod
    def advance(self, x, y):
        return (x + 1, y)

class SOUTH:
    @staticmethod
    def turn_right(self):
        return WEST

    @staticmethod
    def turn_left(self):
        return EAST

    @staticmethod
    def advance(self, x, y):
        return (x, y - 1)

class WEST:
    @staticmethod
    def turn_right(self):
        return NORTH

    @staticmethod
    def turn_left(self):
        return SOUTH

    @staticmethod
    def advance(self, x, y):
        return (x - 1, y)

class Robot:
    def __init__(self, direction=NORTH, x = 0, y = 0):
        self.coordinates = (x, y)
        self.bearing = direction

    def turn_right(self):
        self.bearing = self.bearing.turn_right(self.bearing)

    def turn_left(self):
        self.bearing = self.bearing.turn_left(self.bearing)

    def advance(self):
        self.coordinates = self.bearing.advance(self.bearing, self.coordinates[0], self.coordinates[1])

    def simulate(self, directions):
        for direction in directions:
            if direction == "L":
                self.turn_left()
            elif direction == "R":
                self.turn_right()
            elif direction == "A":
                self.advance()

File: php/robot-simulator/robot-simulator.php
<?php

class Robot {
    const DIRECTION_NORTH = [0, 1];
    const DIRECTION_EAST = [1, 0];
    const DIRECTION_SOUTH = [0, -1];
    const DIRECTION_WEST = [-1, 0];

    function __construct($position, $direction) {
        $this->position = $position;
        $this->direction = $direction;
    }

    function turnRight() {
        switch($this->direction) {
            case self::DIRECTION_NORTH : $this->direction = self::DIRECTION_EAST; break;
            case self::DIRECTION_EAST : $this->direction = self::DIRECTION_SOUTH; break;
            case self::DIRECTION_SOUTH : $this->direction = self::DIRECTION_WEST; break;
            case self::DIRECTION_WEST : $this->direction = self::DIRECTION_NORTH; break;
        }
        return $this;
    }

    function turnLeft() {
        switch($this->direction) {
            case self::DIRECTION_NORTH : $this->direction = self::DIRECTION_WEST; break;
            case self::DIRECTION_EAST : $this->direction = self::DIRECTION_NORTH; break;
            case self::DIRECTION_SOUTH : $this->direction = self::DIRECTION_EAST; break;
            case self::DIRECTION_WEST : $this->direction = self::DIRECTION_SOUTH; break; 
        }
        return $this;
    }

    function advance() {
        $this->position[0] += $this->direction[0];
        $this->position[1] += $this->direction[1];
        return $this;
    }

    function instructions($instructions){
        foreach(str_split($instructions) as $instruction) {
            switch($instruction) {
                case "A" : $this->advance(); break;
                case "L" : $this->turnLeft(); break;
                case "R" : $this->turnRight(); break;
                default : throw new InvalidArgumentException;
            }
        }
    }
}






